[DRAFT] Feature/throws exception

torrancew opened the pull request:
Fixes #1412 


_This is an early draft to solicit feedback. It is functional, but there are a couple of implementation details that felt rather clumsy:_
- _Mapping a user specified qualified name (eg `Xapian::Database::add_document`) to the mangled representations available during function analysis_
- _Representing the transformations themselves within the existing machinery (the throwable bit is more or less just bolted on at this phase)_

_And of course some obvious missing bits like documentation, additional test cases, etc. Let me know what you think about the current approach, and anything you'd like to see improved -- I'm happy to iterate!_

This patch introduces a `throws!` macro, so that individual functions and methods can be marked as throwing an exception. Example usage can be found in the provided unit tests, but generally resembles:

```rust
import_cpp! {
    #include "foo.h"
    generate!("do_something")
    throws!("do_something")
}
```

- [x] CLA signed
- [x] Tests pass
- [ ] Appropriate changes to README are included in PR

adetaylor reviewed (commented):
This feels very much along the right lines. In fact you've gone further than I thought you'd do in the initial iteration by tackling the MoveIt TryNew stuff.

Before I land this, I'd want to see:
* Diagnostics if a `throws!` directive had no effect
* Documentation, obviously
* All existing tests passing (I'll click the Approve and Run button shortly)

Generally speaking this is looking great.

torrancew commented:
Thanks!

The test suite is green locally, and the failing tests appear to be the ones I've added. I'm inclined to suspect a difference in my build environment and CI's. I'll dig into that and try to make them a bit more portable, assuming it's correct. Likewise I'll start in on cleaning up the docs. In the mean time, I added a few more test cases, and in the process uncovered (and fixed) support for constructors as well!

adetaylor commented on engine/src/conversion/analysis/fun/mod.rs:
```diff
@@ -1196,6 +1198,16 @@ impl<'a> FnAnalyzer<'a> {
         }
         let mut cxxbridge_name = make_ident(&cxxbridge_name);
 
+        let friendly_name = {
```
I appreciate that much of the naming stuff is a bit of a mess, but please do your best to be consistent here with the way other configuration directives are approached.

In `byvalue_checker.rs` we have:

```rust
        let pod_requests = config
            .get_pod_requests()
            .iter()
            .map(|ty| QualifiedName::new_from_cpp_name(ty))
            .collect();
```

and the comment in `new_from_cpp_name` suggests this is what you should be using.

adetaylor commented on engine/src/conversion/codegen_rs/fun_codegen.rs:
```diff
@@ -196,6 +198,16 @@ pub(super) fn gen_function(
     let params = unqualify_params(params);
     let ret_type = unqualify_ret_type(ret_type.into_owned());
+
+    let ret_type = if analysis.throwable {
```
Slight preference for a single `match` statement with an arm something like `_ if !analysis.throwable => ret_type`

adetaylor commented:
@torrancew Hi, I've been doing a thorough cleanup of all 27 outstanding autocxx PRs so that we don't get to the point of PR bankruptcy. Yours is the *only* one I haven't ruthlessly closed or merged. It would be good to see it get wrapped up!

torrancew commented:
Thanks for the bump! I'll try to make some time to clean up the diagnostics
and feed back this weekend!

-Tray

On Fri, Feb 28, 2025, 09:15 Adrian Taylor ***@***.***> wrote:

> @torrancew <https://github.com/torrancew> Hi, I've been doing a thorough
> cleanup of all 27 outstanding autocxx PRs so that we don't get to the point
> of PR bankruptcy. Yours is the *only* one I haven't ruthlessly closed or
> merged. It would be good to see it get wrapped up!
>
> —
> Reply to this email directly, view it on GitHub
> <https://github.com/google/autocxx/pull/1426#issuecomment-2691163691>, or
> unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AABZREQYLJ5JCXMAV5TR5OD2SCKT3AVCNFSM6AAAAABVQ7XPBSVHI2DSMVQWIX3LMV43OSLTON2WKQ3PNVWWK3TUHMZDMOJRGE3DGNRZGE>
> .
> You are receiving this because you were mentioned.Message ID:
> ***@***.***>
> [image: adetaylor]*adetaylor* left a comment (google/autocxx#1426)
> <https://github.com/google/autocxx/pull/1426#issuecomment-2691163691>
>
> @torrancew <https://github.com/torrancew> Hi, I've been doing a thorough
> cleanup of all 27 outstanding autocxx PRs so that we don't get to the point
> of PR bankruptcy. Yours is the *only* one I haven't ruthlessly closed or
> merged. It would be good to see it get wrapped up!
>
> —
> Reply to this email directly, view it on GitHub
> <https://github.com/google/autocxx/pull/1426#issuecomment-2691163691>, or
> unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AABZREQYLJ5JCXMAV5TR5OD2SCKT3AVCNFSM6AAAAABVQ7XPBSVHI2DSMVQWIX3LMV43OSLTON2WKQ3PNVWWK3TUHMZDMOJRGE3DGNRZGE>
> .
> You are receiving this because you were mentioned.Message ID:
> ***@***.***>
>
